File: golang/calculator.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

const (
	CHAR_EOF     byte = 0
	CHAR_PLUS    byte = '+'
	CHAR_MINUS   byte = '-'
	CHAR_STAR    byte = '*'
	CHAR_SLASH   byte = '/'
	CHAR_PERCENT byte = '%'
	CHAR_LPAREN  byte = '('
	CHAR_RPAREN  byte = ')'
	CHAR_EQUALS  byte = '='
	CHAR_ZERO    byte = '0'
	CHAR_NINE    byte = '9'
	CHAR_A_LOWER byte = 'a'
	CHAR_Z_LOWER byte = 'z'
	CHAR_A_UPPER byte = 'A'
	CHAR_Z_UPPER byte = 'Z'
	CHAR_SPACE   byte = ' '
	CHAR_TAB     byte = '\t'
	CHAR_NEWLINE byte = '\n'
	CHAR_CR      byte = '\r'
)

type AstNode interface{}
type NumberNode struct{ value int64 }
type VariableNode struct{ name string }
type BinaryOpNode struct {
	op    byte
	left  AstNode
	right AstNode
}
type AssignmentNode struct {
	varName string
	expr    AstNode
}

type Parser struct {
	input   string
	pos     int
	len     int
	current byte
}

func NewParser(input string) *Parser {
	p := &Parser{
		input: input,
		len:   len(input),
	}
	if p.len > 0 {
		p.current = input[0]
	} else {
		p.current = CHAR_EOF
	}
	return p
}

func (p *Parser) parse() []AstNode {
	nodes := make([]AstNode, 0)
	for p.current != CHAR_EOF {
		p.skipWhitespace()
		if p.current == CHAR_EOF {
			break
		}
		node := p.parseExpression()
		if node != nil {
			nodes = append(nodes, node)
		}

		p.skipWhitespace()
		for p.current == CHAR_NEWLINE {
			p.advance()
			p.skipWhitespace()
		}
	}
	return nodes
}

func (p *Parser) parseExpression() AstNode {
	node := p.parseTerm()

	for {
		p.skipWhitespace()

		if p.current == CHAR_PLUS || p.current == CHAR_MINUS {
			op := p.current
			p.advance()
			right := p.parseTerm()
			node = BinaryOpNode{op: op, left: node, right: right}
		} else {
			break
		}
	}

	return node
}

func (p *Parser) parseTerm() AstNode {
	node := p.parseFactor()

	for {
		p.skipWhitespace()

		if p.current == CHAR_STAR || p.current == CHAR_SLASH || p.current == CHAR_PERCENT {
			op := p.current
			p.advance()
			right := p.parseFactor()
			node = BinaryOpNode{op: op, left: node, right: right}
		} else {
			break
		}
	}

	return node
}

func (p *Parser) parseFactor() AstNode {
	p.skipWhitespace()

	if p.isDigit(p.current) {
		return p.parseNumber()
	} else if p.isLetter(p.current) {
		return p.parseVariable()
	} else if p.current == CHAR_LPAREN {
		p.advance()
		node := p.parseExpression()
		p.skipWhitespace()
		if p.current == CHAR_RPAREN {
			p.advance()
		}
		return node
	} else {
		p.advance()
		return NumberNode{value: 0}
	}
}

func (p *Parser) parseNumber() AstNode {
	start := p.pos
	for p.isDigit(p.current) {
		p.advance()
	}
	val, _ := strconv.ParseInt(p.input[start:p.pos], 10, 64)
	return NumberNode{value: val}
}

func (p *Parser) parseVariable() AstNode {
	start := p.pos
	for p.isLetter(p.current) || p.isDigit(p.current) {
		p.advance()
	}
	varName := p.input[start:p.pos]

	p.skipWhitespace()
	if p.current == CHAR_EQUALS {
		p.advance()
		expr := p.parseExpression()
		return AssignmentNode{varName: varName, expr: expr}
	}

	return VariableNode{name: varName}
}

func (p *Parser) advance() {
	p.pos++
	if p.pos >= p.len {
		p.current = CHAR_EOF
	} else {
		p.current = p.input[p.pos]
	}
}

func (p *Parser) skipWhitespace() {
	for p.isWhitespace(p.current) {
		p.advance()
	}
}

func (p *Parser) isDigit(byte byte) bool {
	return byte >= CHAR_ZERO && byte <= CHAR_NINE
}

func (p *Parser) isLetter(byte byte) bool {
	return (byte >= CHAR_A_LOWER && byte <= CHAR_Z_LOWER) ||
		(byte >= CHAR_A_UPPER && byte <= CHAR_Z_UPPER)
}

func (p *Parser) isWhitespace(byte byte) bool {
	return byte == CHAR_SPACE || byte == CHAR_TAB || byte == CHAR_NEWLINE || byte == CHAR_CR
}

type CalculatorAst struct {
	BaseBenchmark
	n           int64
	result      uint32
	text        string
	expressions []AstNode
}

func (c *CalculatorAst) generateRandomProgram(n int) string {
	var builder strings.Builder
	builder.WriteString("v0 = 1\n")

	for i := 0; i < 10; i++ {
		v := i + 1
		builder.WriteString(fmt.Sprintf("v%d = v%d + %d\n", v, v-1, v))
	}

	for i := 0; i < n; i++ {
		v := i + 10
		builder.WriteString(fmt.Sprintf("v%d = v%d + ", v, v-1))

		switch NextInt(10) {
		case 0:
			builder.WriteString(fmt.Sprintf(
				"(v%d / 3) * 4 - %d / (3 + (18 - v%d)) %% v%d + 2 * ((9 - v%d) * (v%d + 7))",
				v-1, i, v-2, v-3, v-6, v-5))
		case 1:
			builder.WriteString(fmt.Sprintf(
				"v%d + (v%d + v%d) * v%d - (v%d / v%d)",
				v-1, v-2, v-3, v-4, v-5, v-6))
		case 2:
			builder.WriteString(fmt.Sprintf("(3789 - (((v%d)))) + 1", v-7))
		case 3:
			builder.WriteString(fmt.Sprintf("4/2 * (1-3) + v%d/v%d", v-9, v-5))
		case 4:
			builder.WriteString(fmt.Sprintf("1+2+3+4+5+6+v%d", v-1))
		case 5:
			builder.WriteString(fmt.Sprintf("(99999 / v%d)", v-3))
		case 6:
			builder.WriteString(fmt.Sprintf("0 + 0 - v%d", v-8))
		case 7:
			builder.WriteString(fmt.Sprintf("((((((((((v%d)))))))))) * 2", v-6))
		case 8:
			builder.WriteString(fmt.Sprintf("%d * (v%d%%6)%%7", i, v-1))
		case 9:
			builder.WriteString(fmt.Sprintf("(1)/(0-v%d) + (v%d)", v-5, v-7))
		}
		builder.WriteString("\n")
	}

	return builder.String()
}

func (c *CalculatorAst) Prepare() {
	c.n = c.ConfigVal("operations")
	c.text = c.generateRandomProgram(int(c.n))
}

func (c *CalculatorAst) Run(iteration_id int) {
	parser := NewParser(c.text)
	c.expressions = parser.parse()
	c.result += uint32(len(c.expressions))
	if len(c.expressions) > 0 {
		lastExpr := c.expressions[len(c.expressions)-1]
		if assign, ok := lastExpr.(AssignmentNode); ok {
			c.result += uint32(Checksum(assign.varName))
		}
	}
}

func (c *CalculatorAst) Checksum() uint32 {
	return c.result
}

func simpleDiv(a, b int64) int64 {
	if b == 0 {
		return 0
	}
	if (a >= 0 && b > 0) || (a < 0 && b < 0) {
		return a / b
	} else {
		return -(abs(a) / abs(b))
	}
}

func simpleMod(a, b int64) int64 {
	if b == 0 {
		return 0
	}
	return a - simpleDiv(a, b)*b
}

func abs(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

type Interpreter struct {
	variables map[string]int64
}

func NewInterpreter() *Interpreter {
	return &Interpreter{
		variables: make(map[string]int64),
	}
}

func (i *Interpreter) evaluate(node AstNode) int64 {
	switch n := node.(type) {
	case NumberNode:
		return n.value
	case VariableNode:
		if val, ok := i.variables[n.name]; ok {
			return val
		}
		return 0
	case BinaryOpNode:
		left := i.evaluate(n.left)
		right := i.evaluate(n.right)

		switch n.op {
		case '+':
			return left + right
		case '-':
			return left - right
		case '*':
			return left * right
		case '/':
			return simpleDiv(left, right)
		case '%':
			return simpleMod(left, right)
		default:
			return 0
		}
	case AssignmentNode:
		value := i.evaluate(n.expr)
		i.variables[n.varName] = value
		return value
	default:
		return 0
	}
}

func (i *Interpreter) run(expressions []AstNode) int64 {
	var result int64 = 0
	for _, expr := range expressions {
		result = i.evaluate(expr)
	}
	return result
}

type CalculatorInterpreter struct {
	BaseBenchmark
	ast    []AstNode
	result uint32
}

func (c *CalculatorInterpreter) Prepare() {
	astBench := &CalculatorAst{BaseBenchmark: BaseBenchmark{className: "Calculator::Interpreter"}}
	astBench.n = c.ConfigVal("operations")
	astBench.Prepare()
	astBench.Run(0)
	c.ast = astBench.expressions
}

func (c *CalculatorInterpreter) Run(iteration_id int) {
	interpreter := NewInterpreter()
	result := interpreter.run(c.ast)
	c.result += uint32(result)
}

func (c *CalculatorInterpreter) Checksum() uint32 {
	return c.result
}

File: fsharp/Calculator.fs
namespace Benchmarks

open System
open System.Text
open System.Collections.Generic

module CharCodes =
    let CHAR_EOF = '\000'
    let CHAR_PLUS = '+'
    let CHAR_MINUS = '-'
    let CHAR_STAR = '*'
    let CHAR_SLASH = '/'
    let CHAR_PERCENT = '%'
    let CHAR_LPAREN = '('
    let CHAR_RPAREN = ')'
    let CHAR_EQUALS = '='
    let CHAR_ZERO = '0'
    let CHAR_NINE = '9'
    let CHAR_A_LOWER = 'a'
    let CHAR_Z_LOWER = 'z'
    let CHAR_A_UPPER = 'A'
    let CHAR_Z_UPPER = 'Z'
    let CHAR_SPACE = ' '
    let CHAR_TAB = '\t'
    let CHAR_NEWLINE = '\n'
    let CHAR_CR = '\r'

[<AbstractClass>]
type Node() = class end

type Number(value: int64) =
    inherit Node()
    member _.Value = value

type Variable(name: string) =
    inherit Node()
    member _.Name = name

type BinaryOp(op: char, left: Node, right: Node) =
    inherit Node()
    member _.Op = op
    member _.Left = left
    member _.Right = right

type Assignment(varName: string, expr: Node) =
    inherit Node()
    member _.Var = varName
    member _.Expr = expr

[<AllowNullLiteral>]
type Parser(input: string) =
    let mutable pos = 0
    let mutable currentChar = if input.Length > 0 then input.[0] else CharCodes.CHAR_EOF
    let expressions = List<Node>()

    let isDigit (char: char) =
        char >= CharCodes.CHAR_ZERO && char <= CharCodes.CHAR_NINE

    let isLetter (char: char) =
        (char >= CharCodes.CHAR_A_LOWER && char <= CharCodes.CHAR_Z_LOWER)
        || (char >= CharCodes.CHAR_A_UPPER && char <= CharCodes.CHAR_Z_UPPER)

    let isWhitespace (char: char) =
        char = CharCodes.CHAR_SPACE
        || char = CharCodes.CHAR_TAB
        || char = CharCodes.CHAR_NEWLINE
        || char = CharCodes.CHAR_CR

    let advance () =
        pos <- pos + 1

        if pos >= input.Length then
            currentChar <- CharCodes.CHAR_EOF
        else
            currentChar <- input.[pos]

    let skipWhitespace () =
        while isWhitespace currentChar do
            advance ()

    let rec parseNumber () =
        let mutable value = 0L

        while isDigit currentChar do
            value <- value * 10L + int64 (currentChar - CharCodes.CHAR_ZERO)
            advance ()

        Number(value) :> Node

    let rec parseVariable () =
        let start = pos

        while isLetter currentChar || isDigit currentChar do
            advance ()

        let varName = input.Substring(start, pos - start)

        skipWhitespace ()

        if currentChar = CharCodes.CHAR_EQUALS then
            advance ()
            let expr = parseExpression ()
            Assignment(varName, expr) :> Node
        else
            Variable(varName) :> Node

    and parseExpression () =
        let mutable node = parseTerm ()
        let mutable continueLoop = true

        while continueLoop do
            skipWhitespace ()

            if currentChar = CharCodes.CHAR_PLUS || currentChar = CharCodes.CHAR_MINUS then
                let op = currentChar
                advance ()
                let right = parseTerm ()
                node <- BinaryOp(op, node, right) :> Node
            else
                continueLoop <- false

        node

    and parseTerm () =
        let mutable node = parseFactor ()
        let mutable continueLoop = true

        while continueLoop do
            skipWhitespace ()

            if
                currentChar = CharCodes.CHAR_STAR
                || currentChar = CharCodes.CHAR_SLASH
                || currentChar = CharCodes.CHAR_PERCENT
            then
                let op = currentChar
                advance ()
                let right = parseFactor ()
                node <- BinaryOp(op, node, right) :> Node
            else
                continueLoop <- false

        node

    and parseFactor () =
        skipWhitespace ()

        if isDigit currentChar then
            parseNumber ()
        elif isLetter currentChar then
            parseVariable ()
        elif currentChar = CharCodes.CHAR_LPAREN then
            advance ()
            let node = parseExpression ()
            skipWhitespace ()

            if currentChar = CharCodes.CHAR_RPAREN then
                advance ()

            node
        else
            advance ()
            Number(0L) :> Node

    member _.Parse() =
        while currentChar <> CharCodes.CHAR_EOF do
            skipWhitespace ()

            if currentChar = CharCodes.CHAR_EOF then
                ()
            else
                let expr = parseExpression ()
                expressions.Add(expr)

                skipWhitespace ()

                while currentChar = CharCodes.CHAR_NEWLINE do
                    advance ()
                    skipWhitespace ()

        List.ofSeq expressions

type Interpreter() =
    let variables = Dictionary<string, int64>()

    let safeAbs (x: int64) =
        if x >= 0L then x
        elif x = Int64.MinValue then Int64.MaxValue
        else -x

    let simpleDiv (a: int64) (b: int64) =
        if b = 0L then
            0L
        elif (a >= 0L && b > 0L) || (a < 0L && b < 0L) then
            a / b
        else
            let absA = safeAbs a
            let absB = safeAbs b
            -(absA / absB)

    let simpleMod (a: int64) (b: int64) =
        if b = 0L then 0L else a - simpleDiv a b * b

    let rec evaluate (node: Node) =
        match node with
        | :? Number as num -> num.Value
        | :? Variable as var ->
            match variables.TryGetValue(var.Name) with
            | true, value -> value
            | false, _ -> 0L
        | :? BinaryOp as bin ->
            let left = evaluate bin.Left
            let right = evaluate bin.Right

            match bin.Op with
            | '+' -> left + right
            | '-' -> left - right
            | '*' -> left * right
            | '/' -> simpleDiv left right
            | '%' -> simpleMod left right
            | _ -> 0L
        | :? Assignment as assign ->
            let value = evaluate assign.Expr
            variables.[assign.Var] <- value
            value
        | _ -> 0L

    member _.Run(expressions: Node list) =
        let mutable result = 0L

        for expr in expressions do
            result <- evaluate expr

        result

module ProgramGenerator =
    let generateRandomProgram (n: int64) =
        let sb = StringBuilder()
        sb.AppendLine("v0 = 1") |> ignore

        for i = 0 to 9 do
            let v = i + 1
            sb.AppendLine($"v{v} = v{v - 1} + {v}") |> ignore

        let mutable i = 0L

        while i < n do
            let v = int (i + 10L)
            sb.Append($"v{v} = v{v - 1} + ") |> ignore

            match Helper.NextInt(10) with
            | 0 ->
                sb.AppendFormat(
                    "(v{0} / 3) * 4 - {1} / (3 + (18 - v{2})) % v{3} + 2 * ((9 - v{4}) * (v{5} + 7))",
                    v - 1,
                    i,
                    v - 2,
                    v - 3,
                    v - 6,
                    v - 5
                )
                |> ignore
            | 1 ->
                sb.AppendFormat("v{0} + (v{1} + v{2}) * v{3} - (v{4} / v{5})", v - 1, v - 2, v - 3, v - 4, v - 5, v - 6)
                |> ignore
            | 2 -> sb.AppendFormat("(3789 - (((v{0})))) + 1", v - 7) |> ignore
            | 3 -> sb.AppendFormat("4/2 * (1-3) + v{0}/v{1}", v - 9, v - 5) |> ignore
            | 4 -> sb.AppendFormat("1+2+3+4+5+6+v{0}", v - 1) |> ignore
            | 5 -> sb.AppendFormat("(99999 / v{0})", v - 3) |> ignore
            | 6 -> sb.AppendFormat("0 + 0 - v{0}", v - 8) |> ignore
            | 7 -> sb.AppendFormat("((((((((((v{0})))))))))) * 2", v - 6) |> ignore
            | 8 -> sb.AppendFormat("{0} * (v{1}%6)%7", i, v - 1) |> ignore
            | 9 -> sb.AppendFormat("(1)/(0-v{0}) + (v{1})", v - 5, v - 7) |> ignore
            | _ -> ()

            sb.AppendLine() |> ignore
            i <- i + 1L

        sb.ToString()

type CalculatorAst() =
    inherit Benchmark()

    let mutable n = 0L
    let mutable text = ""
    let mutable result = 0u
    let mutable expressions: Node list = []

    member this.SetOperations(operations: int64) =
        n <- operations
        text <- ProgramGenerator.generateRandomProgram n

    member this.Parse() =
        let parser = Parser(text)
        expressions <- parser.Parse()

        result <- result + uint32 expressions.Length

        match expressions with
        | [] -> ()
        | _ ->
            let last = expressions |> List.last

            match last with
            | :? Assignment as assign -> result <- result + Helper.Checksum(assign.Var)
            | _ -> ()

    member this.Expressions = expressions

    override this.Checksum = result
    override this.Name = "Calculator::Ast"

    override this.Prepare() =
        n <- this.ConfigVal("operations")
        this.SetOperations(n)
        result <- 0u

    override this.Run(_: int64) = this.Parse()

type CalculatorInterpreter() =
    inherit Benchmark()

    let mutable n = 0L
    let mutable result = 0u
    let mutable ast: Node list = []

    override this.Checksum = result
    override this.Name = "Calculator::Interpreter"

    override this.Prepare() =
        n <- this.ConfigVal("operations")

        let calculator = CalculatorAst()
        calculator.SetOperations(n)
        calculator.Parse()

        ast <- calculator.Expressions

        result <- 0u

    override this.Run(_: int64) =
        let interpreter = Interpreter()
        let calcResult = interpreter.Run(ast)
        result <- result + uint32 calcResult
